File: tests/plugins/flt.lttng-utils.debug-info/test-mip1.cpp
#include "cpp-common/bt2/component-class-dev.hpp"
#include "cpp-common/bt2/graph.hpp"
#include "cpp-common/bt2/plugin-load.hpp"

class MaSource;

class MaIterateur : public bt2::UserMessageIterator<MaIterateur, MaSource>
{
public:
    explicit MaIterateur(const bt2::SelfMessageIterator self, bt2::SelfMessageIteratorConfiguration,
                         bt2::SelfComponentOutputPort) :
        bt2::UserMessageIterator<MaIterateur, MaSource> {self, "MA-ITER"}
    {
    }
    void _next(bt2::ConstMessageArray& msgs);

private:
    bool _mDone = false;

    static void _getSupportedMipVersions(bt2::SelfComponentClass, bt2::ConstValue,
                                         bt2::LoggingLevel, bt2::UnsignedIntegerRangeSet ranges)
    {
        ranges.addRange(0, 1);
    }
};

class MaSource : public bt2::UserSourceComponent<MaSource, MaIterateur>
{
    friend class MaIterateur;

public:
    static constexpr auto name = "ma-source";
    explicit MaSource(const bt2::SelfSourceComponent self, bt2::ConstMapValue, void *) :
        bt2::UserSourceComponent<MaSource, MaIterateur> {self, "MA-SRC"}
    {
        this->_addOutputPort("out");
    }
    static void _getSupportedMipVersions(bt2::SelfComponentClass, bt2::ConstValue,
                                         bt2::LoggingLevel, bt2::UnsignedIntegerRangeSet ranges)
    {
        ranges.addRange(1, 1);
    }
};

void MaIterateur::_next(bt2::ConstMessageArray& msgs)
{
    if (_mDone) {
        return;
    }
    const auto traceCls = this->_selfComponent().createTraceClass();
    const auto trace = traceCls->instantiate();
    trace->nameSpace("le-namespace").name("le-name").uid("le-uid");
    const auto streamCls = traceCls->createStreamClass();
    streamCls->nameSpace("le-namespace")
        .name("le-name")
        .uid("le-uid")
        .defaultClockClass(this->_selfComponent()
                               .createClockClass()
                               ->origin("le-namespace", "le-name", "le-uid")
                               .nameSpace("le-namespace")
                               .name("le-name")
                               .uid("le-uid")
                               .precision(12)
                               .accuracy(24));
    const auto stream = streamCls->instantiate(*trace);
    stream->name("le-name");
    msgs.append(this->_createStreamBeginningMessage(*stream));
    msgs.append(bt2c::call([&] {
        const auto eventMsg = this->_createEventMessage(
            streamCls->createEventClass()
                ->nameSpace("le-namespace")
                .name("le-name")
                .uid("le-uid")
                .payloadFieldClass(
                    (*traceCls->createStructureFieldClass())
                        .appendMember("static-blob",
                                      traceCls->createStaticBlobFieldClass(1)->mediaType(
                                          "application/vnd.rar"))
                        .appendMember(
                            "dynamic-blob-without-field-location",
                            *traceCls->createDynamicBlobWithoutLengthFieldLocationFieldClass())
                        .appendMember("dynamic-blob-with-length-field-location-length",
                                      *traceCls->createUnsignedIntegerFieldClass())
                        .appendMember(
                            "dynamic-blob-with-length-field-location",
                            *traceCls->createDynamicBlobWithLengthFieldLocationFieldClass(
                                *traceCls->createFieldLocation(
                                    bt2::ConstFieldLocation::Scope::EventPayload,
                                    std::vector<const char *> {
                                        "dynamic-blob-with-length-field-location-length"})))
                        .appendMember("bit-array",
                                      traceCls->createBitArrayFieldClass(64)->addFlag(
                                          "flag-foo",
                                          bt2::UnsignedIntegerRangeSet::create()->addRange(0, 12)))
                        .appendMember(
                            "dynamic-array-without-field-location",
                            *traceCls->createDynamicArrayWithoutLengthFieldLocationFieldClass(
                                *traceCls->createUnsignedIntegerFieldClass()))
                        .appendMember("dynamic-array-with-length-field-location-length",
                                      *traceCls->createUnsignedIntegerFieldClass())
                        .appendMember(
                            "dynamic-array-with-length-field-location",
                            *traceCls->createDynamicArrayWithLengthFieldLocationFieldClass(
                                *traceCls->createUnsignedIntegerFieldClass(),
                                *traceCls->createFieldLocation(
                                    bt2::ConstFieldLocation::Scope::EventPayload,
                                    std::vector<const char *> {
                                        "dynamic-array-with-length-field-location-length"})))
                        .appendMember("option-without-selector-field-location",
                                      *traceCls->createOptionWithoutSelectorFieldLocationFieldClass(
                                          *traceCls->createUnsignedIntegerFieldClass()))
                        .appendMember("option-with-bool-selector-field-location-selector",
                                      *traceCls->createBoolFieldClass())
                        .appendMember(
                            "option-with-bool-selector-field-location",
                            *traceCls->createOptionWithBoolSelectorFieldLocationFieldClass(
                                *traceCls->createUnsignedIntegerFieldClass(),
                                *traceCls->createFieldLocation(
                                    bt2::ConstFieldLocation::Scope::EventPayload,
                                    std::vector<const char *> {
                                        "option-with-bool-selector-field-location-selector"})))
                        .appendMember(
                            "option-with-unsigned-integer-selector-field-location-selector",
                            *traceCls->createUnsignedIntegerFieldClass())
                        .appendMember(
                            "option-with-unsigned-integer-selector-field-location",
                            *traceCls->createOptionWithUnsignedIntegerSelectorFieldLocationFieldClass(
                                *traceCls->createUnsignedIntegerFieldClass(),
                                *traceCls->createFieldLocation(
                                    bt2::ConstFieldLocation::Scope::EventPayload,
                                    std::vector<const char *> {
                                        "option-with-unsigned-integer-selector-field-location-selector"}),
                                bt2::UnsignedIntegerRangeSet::create()->addRange(1, 1)))
                        .appendMember("option-with-signed-integer-selector-field-location-selector",
                                      *traceCls->createSignedIntegerFieldClass())
                        .appendMember(
                            "option-with-signed-integer-selector-field-location",
                            *traceCls->createOptionWithSignedIntegerSelectorFieldLocationFieldClass(
                                *traceCls->createUnsignedIntegerFieldClass(),
                                *traceCls->createFieldLocation(
                                    bt2::ConstFieldLocation::Scope::EventPayload,
                                    std::vector<const char *> {
                                        "option-with-signed-integer-selector-field-location-selector"}),
                                bt2::SignedIntegerRangeSet::create()->addRange(1, 1)))
                        .appendMember(
                            "variant-without-selector-field-location",
                            traceCls->createVariantWithoutSelectorFieldLocationFieldClass()
                                ->appendOption("variant-option",
                                               *traceCls->createUnsignedIntegerFieldClass()))
                        .appendMember(
                            "variant-with-unsigned-integer-selector-field-location-selector",
                            *traceCls->createUnsignedIntegerFieldClass())
                        .appendMember(
                            "variant-with-unsigned-integer-selector-field-location",
                            traceCls
                                ->createVariantWithUnsignedIntegerSelectorFieldLocationFieldClass(
                                    *traceCls->createFieldLocation(
                                        bt2::ConstFieldLocation::Scope::EventPayload,
                                        std::vector<const char *> {
                                            "variant-with-unsigned-integer-selector-field-location-selector"}))
                                ->appendOption(
                                    "variant-option", *traceCls->createUnsignedIntegerFieldClass(),
                                    bt2::UnsignedIntegerRangeSet::create()->addRange(1, 1)))
                        .appendMember(
                            "variant-with-signed-integer-selector-field-location-selector",
                            *traceCls->createSignedIntegerFieldClass())
                        .appendMember(
                            "variant-with-signed-integer-selector-field-location",
                            traceCls
                                ->createVariantWithSignedIntegerSelectorFieldLocationFieldClass(
                                    *traceCls->createFieldLocation(
                                        bt2::ConstFieldLocation::Scope::EventPayload,
                                        std::vector<const char *> {
                                            "variant-with-signed-integer-selector-field-location-selector"}))
                                ->appendOption(
                                    "variant-option", *traceCls->createUnsignedIntegerFieldClass(),
                                    bt2::SignedIntegerRangeSet::create()->addRange(1, 1)))),
            *stream, 123);
        const auto payload = *eventMsg->event().payloadField();
        payload["static-blob"]->asBlob().data()[0] = 0x11;
        payload["dynamic-blob-without-field-location"]->asDynamicBlob().length(1).data()[0] = 0x22;
        payload["dynamic-blob-with-length-field-location-length"]->asUnsignedInteger().value(1);
        payload["dynamic-blob-with-length-field-location"]->asDynamicBlob().length(1).data()[0] =
            0x33;
        payload["bit-array"]->asBitArray().valueAsInteger(0xffffffffffffffff);
        payload["dynamic-array-without-field-location"]
            ->asDynamicArray()
            .length(1)[0]
            .asUnsignedInteger()
            .value(1234);
        payload["dynamic-array-with-length-field-location-length"]->asUnsignedInteger().value(1);
        payload["dynamic-array-with-length-field-location"]
            ->asDynamicArray()
            .length(1)[0]
            .asUnsignedInteger()
            .value(2345);
        payload["option-without-selector-field-location"]
            ->asOption()
            .hasField(true)
            .field()
            ->asUnsignedInteger()
            .value(111);
        payload["option-with-bool-selector-field-location-selector"]->asBool().value(1);
        payload["option-with-bool-selector-field-location"]
            ->asOption()
            .hasField(true)
            .field()
            ->asUnsignedInteger()
            .value(222);
        payload["option-with-unsigned-integer-selector-field-location-selector"]
            ->asUnsignedInteger()
            .value(1);
        payload["option-with-unsigned-integer-selector-field-location"]
            ->asOption()
            .hasField(true)
            .field()
            ->asUnsignedInteger()
            .value(333);
        payload["option-with-signed-integer-selector-field-location-selector"]
            ->asSignedInteger()
            .value(1);
        payload["option-with-signed-integer-selector-field-location"]
            ->asOption()
            .hasField(true)
            .field()
            ->asUnsignedInteger()
            .value(333);
        payload["variant-without-selector-field-location"]
            ->asVariant()
            .selectOption(0)
            .selectedOptionField()
            .asUnsignedInteger()
            .value(444);
        payload["variant-with-unsigned-integer-selector-field-location-selector"]
            ->asUnsignedInteger()
            .value(1);
        payload["variant-with-unsigned-integer-selector-field-location"]
            ->asVariant()
            .selectOption(0)
            .selectedOptionField()
            .asUnsignedInteger()
            .value(555);
        payload["variant-with-signed-integer-selector-field-location-selector"]
            ->asSignedInteger()
            .value(1);
        payload["variant-with-signed-integer-selector-field-location"]
            ->asVariant()
            .selectOption(0)
            .selectedOptionField()
            .asUnsignedInteger()
            .value(666);
        return eventMsg;
    }));
    msgs.append(this->_createStreamEndMessage(*stream));
    _mDone = true;
}

int main()
{
    const auto textPlugin = bt2::findPlugin("text");
    BT_ASSERT(textPlugin);
    const auto detailsCompCls = textPlugin->sinkComponentClasses()["details"];
    BT_ASSERT(detailsCompCls);
    const auto lttngUtilsPlugin = bt2::findPlugin("lttng-utils");
    BT_ASSERT(lttngUtilsPlugin);
    const auto debugInfoCompCls = lttngUtilsPlugin->filterComponentClasses()["debug-info"];
    BT_ASSERT(debugInfoCompCls);
    const auto graph = bt2::Graph::create(1);
    const auto maSourceCompCls = bt2::SourceComponentClass::create<MaSource>();
    const auto src = graph->addComponent(*maSourceCompCls, "la-source");
    const auto snk = graph->addComponent(*detailsCompCls, "le-sink");

    /* For easy flipping between using the filter and not */
    const bool runWithFilter = true;
    if (runWithFilter) {
        const auto flt __attribute__((unused)) = graph->addComponent(*debugInfoCompCls, "le-filtre");

        graph->connectPorts(*src.outputPorts()["out"], *flt.inputPorts()["in"]);
        graph->connectPorts(*flt.outputPorts()["out"], *snk.inputPorts()["in"]);
    } else {
        graph->connectPorts(*src.outputPorts()["out"], *snk.inputPorts()["in"]);
    }

    graph->run();
}

File: tests/bindings/python/bt2/test_mip1.py
import typing
import unittest

import bt2

# pyright: strict,  reportPrivateUsage=false


T = typing.TypeVar("T")


# Assert that `obj` is of type `expected_type` and return it as a `T`.
def c(obj: object, expected_type: typing.Type[T]) -> T:
    assert type(obj) is expected_type
    return obj


class TheSourceIterator(bt2._UserMessageIterator):
    def __init__(
        self,
        config: bt2._MessageIteratorConfiguration,
        port: bt2._UserComponentOutputPort,
    ):
        tc = self._component._create_trace_class()
        t = tc(
            namespace="the-trace-namespace", name="the-trace-name", uid="the-trace-uid"
        )

        cc = self._component._create_clock_class(
            origin=bt2.ClockClassOrigin(
                "the-clock-class-origin-namespace",
                "the-clock-class-origin-name",
                "the-clock-class-origin-uid",
            ),
            namespace="the-clock-class-namespace",
            name="the-clock-class-name",
            uid="the-clock-class-uid",
            precision=12,
            accuracy=34,
        )

        sc = tc.create_stream_class(
            namespace="the-stream-class-namespace",
            name="the-stream-class-name",
            uid="the-stream-class-uid",
            default_clock_class=cc,
        )
        s = t.create_stream(sc)

        ec = sc.create_event_class(
            namespace="the-event-class-namespace",
            name="the-event-class-name",
            uid="the-event-class-uid",
            payload_field_class=self._create_payload_fc(tc),
        )
        em = self._create_event_message(ec, s, default_clock_snapshot=100)
        e = em.event
        p = e.payload_field
        assert p is not None
        self._fill_payload_field(p)

        self._msgs = [
            self._create_stream_beginning_message(s),
            em,
            self._create_stream_end_message(s),
        ]

    def _create_payload_fc(self, tc: bt2._TraceClass):
        payload_fc = tc.create_structure_field_class()

        # Static blob
        payload_fc.append_member(
            "static-blob",
            tc.create_static_blob_field_class(4, media_type="application/vnd.rar"),
        )

        # Dynamic blob witout length field
        payload_fc.append_member(
            "dynamic-blob-without-length-field",
            tc.create_dynamic_blob_field_class(),
        )

        # Dynamic blob with length field
        payload_fc.append_member(
            "dynamic-blob-with-length-field-length",
            tc.create_unsigned_integer_field_class(),
        )
        payload_fc.append_member(
            "dynamic-blob-with-length-field",
            tc.create_dynamic_blob_field_class(
                length_field_location=tc.create_field_location(
                    bt2.FieldLocationScope.EVENT_PAYLOAD,
                    ["dynamic-blob-with-length-field-length"],
                )
            ),
        )

        # Bit array (with flags)
        bit_array_fc = tc.create_bit_array_field_class(16)
        bit_array_fc.add_flag("flag1", bt2.UnsignedIntegerRangeSet([0]))
        bit_array_fc += [
            ("flag2", bt2.UnsignedIntegerRangeSet([4, (12, 15)])),
            ("flag3", bt2.UnsignedIntegerRangeSet([(4, 7)])),
        ]
        payload_fc.append_member("bit-array", bit_array_fc)

        # Option without selector field
        payload_fc.append_member(
            "option-without-selector-field",
            tc.create_option_without_selector_field_class(
                tc.create_string_field_class()
            ),
        )

        # Option with bool selector field
        payload_fc.append_member(
            "option-with-bool-selector-field-selector", tc.create_bool_field_class()
        )
        payload_fc.append_member(
            "option-with-bool-selector-field",
            tc.create_option_with_bool_selector_field_class(
                tc.create_string_field_class(),
                selector_field_location=tc.create_field_location(
                    bt2.FieldLocationScope.EVENT_PAYLOAD,
                    ["option-with-bool-selector-field-selector"],
                ),
            ),
        )

        # Option with unsigned integer selector field
        payload_fc.append_member(
            "option-with-unsigned-integer-selector-field-selector",
            tc.create_unsigned_integer_field_class(8),
        )
        payload_fc.append_member(
            "option-with-unsigned-integer-selector-field",
            tc.create_option_with_unsigned_integer_selector_field_class(
                tc.create_string_field_class(),
                selector_field_location=tc.create_field_location(
                    bt2.FieldLocationScope.EVENT_PAYLOAD,
                    ["option-with-unsigned-integer-selector-field-selector"],
                ),
                ranges=bt2.UnsignedIntegerRangeSet([1]),
            ),
        )

        # Option with signed integer selector field
        payload_fc.append_member(
            "option-with-signed-integer-selector-field-selector",
            tc.create_signed_integer_field_class(8),
        )
        payload_fc.append_member(
            "option-with-signed-integer-selector-field",
            tc.create_option_with_signed_integer_selector_field_class(
                tc.create_string_field_class(),
                selector_field_location=tc.create_field_location(
                    bt2.FieldLocationScope.EVENT_PAYLOAD,
                    ["option-with-signed-integer-selector-field-selector"],
                ),
                ranges=bt2.SignedIntegerRangeSet([1]),
            ),
        )

        # Variant without selector field location
        variant_without_selector_field_fc = (
            tc.create_variant_without_selector_field_class()
        )
        variant_without_selector_field_fc.append_option(
            None, tc.create_single_precision_real_field_class()
        )
        variant_without_selector_field_fc.append_option(
            "str", tc.create_string_field_class()
        )
        payload_fc.append_member(
            "variant-without-selector-field",
            variant_without_selector_field_fc,
        )

        # Variant with unsigned integer selector field location
        payload_fc.append_member(
            "variant-with-unsigned-integer-selector-field-selector",
            tc.create_unsigned_integer_field_class(8),
        )
        variant_with_unsigned_integer_selector_field_fc = (
            tc.create_variant_with_unsigned_selector_field_class(
                selector_field_location=tc.create_field_location(
                    bt2.FieldLocationScope.EVENT_PAYLOAD,
                    ["variant-with-unsigned-integer-selector-field-selector"],
                ),
            )
        )
        variant_with_unsigned_integer_selector_field_fc.append_option(
            None,
            tc.create_single_precision_real_field_class(),
            bt2.UnsignedIntegerRangeSet([0]),
        )
        variant_with_unsigned_integer_selector_field_fc.append_option(
            "str", tc.create_string_field_class(), bt2.UnsignedIntegerRangeSet([1])
        )
        payload_fc.append_member(
            "variant-with-unsigned-integer-selector-field",
            variant_with_unsigned_integer_selector_field_fc,
        )

        # Variant with signed integer selector field location
        payload_fc.append_member(
            "variant-with-signed-integer-selector-field-selector",
            tc.create_signed_integer_field_class(8),
        )
        variant_with_signed_integer_selector_field_fc = (
            tc.create_variant_with_signed_selector_field_class(
                selector_field_location=tc.create_field_location(
                    bt2.FieldLocationScope.EVENT_PAYLOAD,
                    ["variant-with-signed-integer-selector-field-selector"],
                ),
            )
        )
        variant_with_signed_integer_selector_field_fc.append_option(
            None,
            tc.create_single_precision_real_field_class(),
            bt2.SignedIntegerRangeSet([0]),
        )
        variant_with_signed_integer_selector_field_fc.append_option(
            "str", tc.create_string_field_class(), bt2.SignedIntegerRangeSet([1])
        )
        payload_fc.append_member(
            "variant-with-signed-integer-selector-field",
            variant_with_signed_integer_selector_field_fc,
        )

        return payload_fc

    def _fill_payload_field(self, p: bt2._StructureField):
        # Static blob
        static_blob = c(p["static-blob"], bt2._StaticBlobField)
        static_blob.data = b"abcd"
        static_blob.data[2] = 0x65  # e

        # Dynamic blob without length field
        dynamic_blob_without_length_field = c(
            p["dynamic-blob-without-length-field"], bt2._DynamicBlobField
        )
        dynamic_blob_without_length_field.length = 5
        dynamic_blob_without_length_field.data = b"salut"

        # Dynamic blob with length field
        c(
            p["dynamic-blob-with-length-field-length"],
            bt2._UnsignedIntegerField,
        ).value = 7
        dynamic_blob_with_length_field = c(
            p["dynamic-blob-with-length-field"],
            bt2._DynamicBlobWithLengthFieldField,
        )
        dynamic_blob_with_length_field.length = 7
        dynamic_blob_with_length_field.data = b"bonjour"

        # Bit array
        c(
            p["bit-array"],
            bt2._BitArrayField,
        ).value_as_integer = 0b10000

        # Option without selector field
        c(
            p["option-without-selector-field"],
            bt2._OptionField,
        ).value = "hi"

        # Option with bool selector field
        c(
            p["option-with-bool-selector-field-selector"],
            bt2._BoolField,
        ).value = True
        c(
            p["option-with-bool-selector-field"],
            bt2._OptionWithBoolSelectorField,
        ).value = "hello"

        # Option with unsigned integer selector field
        c(
            p["option-with-unsigned-integer-selector-field-selector"],
            bt2._UnsignedIntegerField,
        ).value = 1
        c(
            p["option-with-unsigned-integer-selector-field"],
            bt2._OptionWithUnsignedIntegerSelectorField,
        ).value = "world"

        # Option with signed integer selector field
        c(
            p["option-with-signed-integer-selector-field-selector"],
            bt2._SignedIntegerField,
        ).value = 1
        c(
            p["option-with-signed-integer-selector-field"],
            bt2._OptionWithSignedIntegerSelectorField,
        ).value = "option-with-signed-integer-selector-field-value"

        # Variant without selector field
        c(
            p["variant-without-selector-field"],
            bt2._VariantField,
        ).selected_option_index = 1
        c(
            p["variant-without-selector-field"],
            bt2._VariantField,
        ).value = "variant-without-selector-field-value"

        # Variant with unsigned integer selector field
        c(
            p["variant-with-unsigned-integer-selector-field-selector"],
            bt2._UnsignedIntegerField,
        ).value = 1
        variant_field_with_unsigned_integer_selector_field = c(
            p["variant-with-unsigned-integer-selector-field"],
            bt2._VariantFieldWithUnsignedIntegerSelector,
        )
        variant_field_with_unsigned_integer_selector_field.selected_option_index = 1
        variant_field_with_unsigned_integer_selector_field.value = (
            "variant-with-unsigned-integer-selector-field-value"
        )

        # Variant with signed integer selector field
        c(
            p["variant-with-signed-integer-selector-field-selector"],
            bt2._SignedIntegerField,
        ).value = 1
        variant_field_with_signed_integer_selector_field = c(
            p["variant-with-signed-integer-selector-field"],
            bt2._VariantFieldWithSignedIntegerSelector,
        )
        variant_field_with_signed_integer_selector_field.selected_option_index = 1
        variant_field_with_signed_integer_selector_field.value = (
            "variant-with-signed-integer-selector-field-value"
        )

    def __next__(self):
        if len(self._msgs) == 0:
            raise StopIteration

        return self._msgs.pop(0)


class TheSource(bt2._UserSourceComponent, message_iterator_class=TheSourceIterator):
    def __init__(
        self,
        config: bt2._UserSourceComponentConfiguration,
        params: bt2._MapValueConst,
        obj: object,
    ):
        self._add_output_port("out")


class TheSink(bt2._UserSinkComponent):
    def __init__(
        self,
        config: bt2._UserSinkComponentConfiguration,
        params: bt2._MapValueConst,
        obj: object,
    ):
        self._in = self._add_input_port("in")
        self._test = typing.cast(unittest.TestCase, obj)

    def _user_graph_is_configured(self):
        self._iter = self._create_message_iterator(self._in)

    def _test_blob(
        self,
        static_blob: bt2._StaticBlobFieldConst,
        dynamic_blob_without_length_field: bt2._DynamicBlobFieldConst,
        dynamic_blob_with_length_field: bt2._DynamicBlobWithLengthFieldFieldConst,
    ):
        # Test `_BlobFieldClassConst.media_type`
        self._test.assertEqual(static_blob.cls.media_type, "application/vnd.rar")

        # Test `_StaticBlobFieldClassConst.length`
        self._test.assertEqual(static_blob.cls.length, 4)

        # Test `_DynamicBlobWithLengthFieldFieldClassConst.length_field_location`
        fl = dynamic_blob_with_length_field.cls.length_field_location
        self._test.assertEqual(fl.root_scope, bt2.FieldLocationScope.EVENT_PAYLOAD)

        # Test `_FieldLocationConst.__len__`
        self._test.assertEqual(len(fl), 1)

        # Test `_FieldLocationConst.__iter__`
        self._test.assertEqual(list(fl), ["dynamic-blob-with-length-field-length"])

        self._test.assertEqual(static_blob.data, b"abed")
        self._test.assertEqual(len(static_blob), 4)

        self._test.assertEqual(dynamic_blob_without_length_field.data, b"salut")
        self._test.assertEqual(dynamic_blob_without_length_field.length, 5)

    def _test_bit_array(self, bit_array: bt2._BitArrayFieldConst):
        bit_array_fc = bit_array.cls
        self._test.assertIsInstance(bit_array_fc, bt2._BitArrayFieldClassConst)

        # Test `_BitArrayFieldClassConst.length`
        self._test.assertEqual(bit_array_fc.length, 16)

        # Test `_BitArrayFieldClassConst.active_flag_labels_for_value_as_integer`
        self._test.assertEqual(
            bit_array_fc.active_flag_labels_for_value_as_integer(0b1), ["flag1"]
        )
        self._test.assertEqual(
            bit_array_fc.active_flag_labels_for_value_as_integer(0b100000000), []
        )

        # Test `_BitArrayFieldClassConst.flags`
        flags = bit_array_fc.flags
        self._test.assertIsInstance(flags, bt2._BitArrayFlags)

        # Test `_BitArrayFlags.__len__`
        self._test.assertEqual(len(flags), 3)

        def flag_to_tuple(flag: bt2._BitArrayFlag):
            self._test.assertIsInstance(flag.label, str)
            self._test.assertIsInstance(flag.ranges, bt2._UnsignedIntegerRangeSetConst)
            return (flag.label, [(rng.lower, rng.upper) for rng in flag.ranges])

        def flags_to_list(flags: bt2._BitArrayFlags):
            return [flag_to_tuple(flag) for flag in flags]

        # Test `_BitArrayFlags.__iter__`, `_BitArrayFlag.label` and
        # `_BitArrayFlag.ranges`
        self._test.assertEqual(
            flags_to_list(flags),
            [
                ("flag1", [(0, 0)]),
                ("flag2", [(4, 4), (12, 15)]),
                ("flag3", [(4, 7)]),
            ],
        )

        # Test `_BitArrayFlags.__getitem__`
        self._test.assertEqual(flag_to_tuple(flags["flag1"]), ("flag1", [(0, 0)]))
        self._test.assertEqual(
            flag_to_tuple(flags["flag2"]), ("flag2", [(4, 4), (12, 15)])
        )
        self._test.assertEqual(flag_to_tuple(flags["flag3"]), ("flag3", [(4, 7)]))

        # Test `_BitArrayFieldConst.active_flag_labels`
        self._test.assertEqual(bit_array.active_flag_labels, ["flag2", "flag3"])

    def _test_option_with_bool_selector_field(
        self, field: bt2._OptionWithBoolSelectorFieldConst
    ):
        sfl = field.cls.selector_field_location
        self._test.assertEqual(sfl.root_scope, bt2.FieldLocationScope.EVENT_PAYLOAD)
        self._test.assertEqual(list(sfl), ["option-with-bool-selector-field-selector"])

    def _test_option_with_unsigned_integer_selector_field(
        self, field: bt2._OptionWithUnsignedIntegerSelectorFieldConst
    ):
        sfl = field.cls.selector_field_location
        self._test.assertEqual(sfl.root_scope, bt2.FieldLocationScope.EVENT_PAYLOAD)
        self._test.assertEqual(
            list(sfl), ["option-with-unsigned-integer-selector-field-selector"]
        )

    def _test_option_with_signed_integer_selector_field(
        self, field: bt2._OptionWithSignedIntegerSelectorFieldConst
    ):
        sfl = field.cls.selector_field_location
        self._test.assertEqual(sfl.root_scope, bt2.FieldLocationScope.EVENT_PAYLOAD)
        self._test.assertEqual(
            list(sfl), ["option-with-signed-integer-selector-field-selector"]
        )

    def _test_variant_without_selector_field(self, field: bt2._VariantFieldConst):
        self._test.assertEqual(field.selected_option_index, 1)
        self._test.assertIs(type(field.selected_option), bt2._StringFieldConst)
        self._test.assertEqual(
            field.selected_option, "variant-without-selector-field-value"
        )
        self._test.assertEqual(list(field.cls), [None, "str"])
        self._test.assertIsNone(field.cls.option_at_index(0).name, None)
        self._test.assertEqual(field.cls.option_at_index(1).name, "str")

    def _test_variant_with_unsigned_integer_selector_field(
        self, field: bt2._VariantField
    ):
        self._test.assertEqual(field.selected_option_index, 1)

    def _user_consume(self):
        msg = next(self._iter)

        if type(msg) is bt2._StreamBeginningMessageConst:
            s = msg.stream
            t = s.trace
            sc = s.cls

            self._test.assertEqual(t.namespace, "the-trace-namespace")
            self._test.assertEqual(t.name, "the-trace-name")
            self._test.assertEqual(t.uid, "the-trace-uid")

            cc = sc.default_clock_class
            assert cc is not None
            self._test.assertEqual(cc.namespace, "the-clock-class-namespace")
            self._test.assertEqual(cc.name, "the-clock-class-name")
            self._test.assertEqual(cc.uid, "the-clock-class-uid")
            self._test.assertEqual(cc.opt_precision, 12)
            self._test.assertEqual(cc.accuracy, 34)

            origin = cc.origin
            self._test.assertEqual(origin.namespace, "the-clock-class-origin-namespace")
            self._test.assertEqual(origin.name, "the-clock-class-origin-name")
            self._test.assertEqual(origin.uid, "the-clock-class-origin-uid")

            sc = msg.stream.cls
            self._test.assertEqual(sc.namespace, "the-stream-class-namespace")
            self._test.assertEqual(sc.name, "the-stream-class-name")
            self._test.assertEqual(sc.uid, "the-stream-class-uid")

            ec = sc[0]
            self._test.assertEqual(ec.namespace, "the-event-class-namespace")
            self._test.assertEqual(ec.name, "the-event-class-name")
            self._test.assertEqual(ec.uid, "the-event-class-uid")
        elif type(msg) is bt2._EventMessageConst:
            p = msg.event.payload_field
            assert p is not None

            self._test_blob(
                c(p["static-blob"], bt2._StaticBlobFieldConst),
                c(
                    p["dynamic-blob-without-length-field"],
                    bt2._DynamicBlobFieldConst,
                ),
                c(
                    p["dynamic-blob-with-length-field"],
                    bt2._DynamicBlobWithLengthFieldFieldConst,
                ),
            )
            self._test_bit_array(c(p["bit-array"], bt2._BitArrayFieldConst))
            self._test_option_with_bool_selector_field(
                c(
                    p["option-with-bool-selector-field"],
                    bt2._OptionWithBoolSelectorFieldConst,
                )
            )
            self._test_option_with_unsigned_integer_selector_field(
                c(
                    p["option-with-unsigned-integer-selector-field"],
                    bt2._OptionWithUnsignedIntegerSelectorFieldConst,
                )
            )
            self._test_option_with_signed_integer_selector_field(
                c(
                    p["option-with-signed-integer-selector-field"],
                    bt2._OptionWithSignedIntegerSelectorFieldConst,
                )
            )
            self._test_variant_without_selector_field(
                c(p["variant-without-selector-field"], bt2._VariantFieldConst)
            )


class TestMIP1(unittest.TestCase):
    def test_everything(self):
        g = bt2.Graph(1)
        src = g.add_component(TheSource, "the-source")
        snk = g.add_component(TheSink, "the-sink", None, self)
        g.connect_ports(src.output_ports["out"], snk.input_ports["in"])
        g.run()


if __name__ == "__main__":
    unittest.main()
